![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2FApplied+GenAI%2FSolutions%2Fmulti-type-parsing-and-vector-search-2&file=search_and_query.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%20GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/Applied%20GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2FApplied%2520GenAI%2FSolutions%2Fmulti-type-parsing-and-vector-search-2%2Fsearch_and_query.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%20GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/Applied%2520GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/Applied%20GenAI/Solutions/multi-type-parsing-and-vector-search-2/search_and_query.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Search & Query: Vector Search 2.0 Collections

This notebook is organized by **search taxonomy** — each section demonstrates a search method, making it a reference for "how do I do X?" All demonstrations use the **Combined** collection (all source types in one index) unless noted otherwise.

**Search Taxonomy**

| Method | API | What Drives Results | Vectors Supplied |
|--------|-----|---------------------|------------------|
| **Query** | `query_data_objects` | Field filters (`$eq`, `$gt`, `$in`) | None |
| **SemanticSearch** | `search_data_objects` | Auto-embedded dense similarity | None (auto) |
| **TextSearch** | `search_data_objects` | Native keyword matching | None |
| **VectorSearch (dense)** | `search_data_objects` | Dense vector you supply | Dense |
| **VectorSearch (sparse)** | `search_data_objects` | Sparse vector you supply (BM25) | Sparse |
| **VectorSearch (dense+sparse)** | `batch_search_data_objects` | Two VectorSearch channels via RRF | Dense + Sparse |
| **Hybrid: Semantic+Text (RRF)** | `batch_search_data_objects` | RRF fusion | None (auto) |
| **Hybrid: Semantic+Sparse (RRF)** | `batch_search_data_objects` | RRF fusion | Sparse |
| **Hybrid: 3-way (RRF)** | `batch_search_data_objects` | 3-channel RRF | Sparse |
| **Reranked: VertexRanker** | `batch_search_data_objects` | Cross-encoder reranking (not yet supported) | Varies |
| **Crowding** | Async parallel filtered search | Per-group diversity | None (auto) |

**Key distinctions:**
- **Query vs Search**: Query = metadata filtering only, no relevance ranking (like SQL `WHERE`). Search = ranked results by relevance.
- **Semantic vs Vector**: SemanticSearch = you provide text, VS2 auto-embeds it. VectorSearch = you provide raw vector(s) — required for sparse search and external embedding models.
- **Dense + Sparse**: VectorSearch uses a `oneof` for `vector`/`sparse_vector` — you can supply one per channel. To combine both, use `batch_search_data_objects` with two VectorSearch channels fused via RRF.
- **SearchHint (ANN vs kNN)**: `VectorSearch` and `SemanticSearch` accept a `search_hint` parameter to control index usage. By default, VS2 auto-selects an ANN index if one exists, falling back to brute-force kNN. Use `SearchHint(use_knn=True)` to force exact kNN search (useful for recall evaluation). `TextSearch` does not support `search_hint`. See step 10 for ANN index management and recall benchmarking.

**Embedding task types:** The collections were indexed with `task_type="RETRIEVAL_DOCUMENT"` (see step 5). At search time, we use `task_type="QUESTION_ANSWERING"` — the asymmetric counterpart that tells the embedding model the search text is a question looking for answers in the indexed documents. See [embedding task types](https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/task-types).

---
# Setup

In [1]:
import string
from collections import Counter

import asyncio
import nltk
import numpy as np
import pandas as pd
from rank_bm25 import BM25Okapi
from google.cloud import bigquery, vectorsearch_v1beta
from google import genai
from config import (
    PROJECT_ID, REGION, EMBEDDING_MODEL, EMBEDDING_DIMENSIONS,
    BQ_PROJECT, BQ_DATASET, BQ_TABLE_PREFIX,
    VS_COLLECTION_REDDIT, VS_COLLECTION_ZOOM, VS_COLLECTION_PDF, VS_COLLECTION_COMBINED,
    BM25_K1, BM25_B, BM25_EPSILON, BM25_NGRAMS,
    ANN_INDEX_ID,
)

# NLTK resources for BM25 preprocessing
nltk.download('stopwords', quiet=True)
nltk.download('punkt_tab', quiet=True)
nltk.download('wordnet', quiet=True)
lemmatizer = nltk.stem.WordNetLemmatizer()
stop_words = set(nltk.corpus.stopwords.words('english'))

# Sync clients
vs_client = vectorsearch_v1beta.VectorSearchServiceClient()
do_client = vectorsearch_v1beta.DataObjectServiceClient()
search_client = vectorsearch_v1beta.DataObjectSearchServiceClient()

# Async client for parallel searches (used in crowding/diversity queries)
async_search_client = vectorsearch_v1beta.DataObjectSearchServiceAsyncClient()

# BigQuery + Gemini clients
bq = bigquery.Client(project=BQ_PROJECT)
genai_client = genai.Client(vertexai=True, project=PROJECT_ID, location=REGION)

# Collection parent paths
REDDIT = f'projects/{PROJECT_ID}/locations/{REGION}/collections/{VS_COLLECTION_REDDIT}'
ZOOM = f'projects/{PROJECT_ID}/locations/{REGION}/collections/{VS_COLLECTION_ZOOM}'
PDF = f'projects/{PROJECT_ID}/locations/{REGION}/collections/{VS_COLLECTION_PDF}'
COMBINED = f'projects/{PROJECT_ID}/locations/{REGION}/collections/{VS_COLLECTION_COMBINED}'

print(f'Reddit:   {VS_COLLECTION_REDDIT}')
print(f'Zoom:     {VS_COLLECTION_ZOOM}')
print(f'PDF:      {VS_COLLECTION_PDF}')
print(f'Combined: {VS_COLLECTION_COMBINED}')
print(f'Embedding: {EMBEDDING_MODEL} @ {EMBEDDING_DIMENSIONS} dims')

Reddit:   applied-genai-solutions-multi-type-parsing-reddit
Zoom:     applied-genai-solutions-multi-type-parsing-zoom
PDF:      applied-genai-solutions-multi-type-parsing-pdf
Combined: applied-genai-solutions-multi-type-parsing-combined
Embedding: gemini-embedding-001 @ 768 dims


In [2]:
def search_df(results, fields, max_text=150):
    """Format search results as a ranked DataFrame."""
    rows = []
    for r in results:
        data = r.data_object.data
        row = {}
        for f in fields:
            val = data.get(f, '')
            if isinstance(val, str) and len(val) > max_text:
                val = val[:max_text] + '...'
            elif hasattr(val, '__iter__') and not isinstance(val, str):
                val = ', '.join(str(v) for v in val)
            row[f] = val
        rows.append(row)
    df = pd.DataFrame(rows)
    df.index = range(1, len(df) + 1)
    df.index.name = 'rank'
    return df


def query_df(results, fields, max_text=150):
    """Format query results as a DataFrame."""
    rows = []
    for r in results:
        row = {}
        for f in fields:
            val = r.data.get(f, '')
            if isinstance(val, str) and len(val) > max_text:
                val = val[:max_text] + '...'
            elif hasattr(val, '__iter__') and not isinstance(val, str):
                val = ', '.join(str(v) for v in val)
            row[f] = val
        rows.append(row)
    return pd.DataFrame(rows)


def hybrid_search(parent, query, weights, top_k=10, fields=None):
    """Run hybrid semantic + text search with RRF reranking."""
    if fields is None:
        fields = ['chunk_id', 'text_content']
    output = vectorsearch_v1beta.OutputFields(data_fields=fields)
    request = vectorsearch_v1beta.BatchSearchDataObjectsRequest(
        parent=parent,
        searches=[
            vectorsearch_v1beta.Search(
                semantic_search=vectorsearch_v1beta.SemanticSearch(
                    search_text=query,
                    search_field='text_content_embedding',
                    task_type='QUESTION_ANSWERING',
                    top_k=top_k,
                    output_fields=output,
                )
            ),
            vectorsearch_v1beta.Search(
                text_search=vectorsearch_v1beta.TextSearch(
                    search_text=query,
                    data_field_names=['text_content'],
                    top_k=top_k,
                    output_fields=output,
                )
            ),
        ],
        combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
            ranker=vectorsearch_v1beta.Ranker(
                rrf=vectorsearch_v1beta.ReciprocalRankFusion(weights=weights)
            )
        ),
    )
    response = search_client.batch_search_data_objects(request)
    if response.results:
        return list(response.results[0].results)
    return []


async def crowded_search(parent, query, crowd_field, crowd_values, top_k_per_value=2, fields=None):
    """Diversity search: top_k results per unique value of a metadata field, run in parallel."""
    if fields is None:
        fields = ['chunk_id', 'source_type', 'text_content']
    output = vectorsearch_v1beta.OutputFields(data_fields=fields)

    async def _search_one(value):
        response = await async_search_client.search_data_objects(
            vectorsearch_v1beta.SearchDataObjectsRequest(
                parent=parent,
                semantic_search=vectorsearch_v1beta.SemanticSearch(
                    search_text=query,
                    search_field='text_content_embedding',
                    task_type='QUESTION_ANSWERING',
                    top_k=top_k_per_value,
                    output_fields=output,
                    filter={crowd_field: {'$eq': value}},
                ),
            )
        )
        return value, list(response.results)

    return dict(await asyncio.gather(*[_search_one(v) for v in crowd_values]))


def crowded_df(grouped_results, fields, max_text=150):
    """Format crowded search results as a DataFrame with group labels."""
    rows = []
    for group, results in grouped_results.items():
        for rank, r in enumerate(results, 1):
            data = r.data_object.data
            row = {'group': group, 'group_rank': rank}
            for f in fields:
                val = data.get(f, '')
                if isinstance(val, str) and len(val) > max_text:
                    val = val[:max_text] + '...'
                elif hasattr(val, '__iter__') and not isinstance(val, str):
                    val = ', '.join(str(v) for v in val)
                row[f] = val
            rows.append(row)
    return pd.DataFrame(rows)

### BM25 Helpers

Load the BM25 vocabulary from BigQuery (stored by `build_bm25.py` in step 7a) and define functions for computing sparse embeddings at query time. The vocabulary and model parameters are read from BQ — the same approach used by `apply_bm25.py` (step 7b) to reconstruct the scoring function without needing the `rank_bm25` library.

In [3]:
# Load BM25 vocabulary and model metadata from BigQuery
vocab_table = f'{BQ_PROJECT}.{BQ_DATASET}.{BQ_TABLE_PREFIX}_bm25_vocabulary'
model_table = f'{BQ_PROJECT}.{BQ_DATASET}.{BQ_TABLE_PREFIX}_bm25_model'

vocab_rows = list(bq.query(f'SELECT term, term_index, idf FROM `{vocab_table}`').result())
vocab_map = {row.term: row.term_index for row in vocab_rows}
idf_map = {row.term: row.idf for row in vocab_rows}

model_meta = list(bq.query(f'SELECT * FROM `{model_table}` ORDER BY model_version DESC LIMIT 1').result())[0]

print(f'BM25 model v{model_meta.model_version}: vocab={model_meta.vocab_size}, corpus={model_meta.corpus_size}, avgdl={model_meta.avgdl:.1f}')
print(f'  k1={model_meta.k1}, b={model_meta.b}, epsilon={model_meta.epsilon}, ngrams={model_meta.ngrams}')


def chunk_prep(text, ngrams=BM25_NGRAMS):
    """Preprocess text for BM25: lowercase, remove punctuation, tokenize,
    lemmatize, remove stopwords, generate n-grams."""
    text = text.lower().translate(str.maketrans('', '', string.punctuation))
    tokens = nltk.tokenize.word_tokenize(text)
    unigrams = [lemmatizer.lemmatize(t) for t in tokens
                if len(t) > 1 and t not in stop_words]
    if ngrams >= 2:
        all_grams = unigrams.copy()
        for n in range(2, ngrams + 1):
            all_grams.extend(' '.join(g) for g in zip(*[unigrams[i:] for i in range(n)]))
        return all_grams
    return unigrams


def create_bm25_sparse_embedding(text):
    """Compute BM25 sparse embedding for query text using stored vocabulary."""
    tokens = chunk_prep(text)
    term_freqs = Counter(tokens)
    doc_len = len(tokens)
    k1, b, avgdl = model_meta.k1, model_meta.b, model_meta.avgdl

    indices, values = [], []
    for term, freq in term_freqs.items():
        if term not in vocab_map:
            continue
        idf = idf_map[term]
        score = idf * (freq * (k1 + 1)) / (freq + k1 * (1 - b + b * doc_len / avgdl))
        indices.append(vocab_map[term])
        values.append(score)
    return indices, values

BM25 model v1: vocab=13506, corpus=117, avgdl=407.8
  k1=1.2, b=0.6, epsilon=0.25, ngrams=2


### Collection Counts

In [4]:
for label, parent in [('Reddit', REDDIT), ('Zoom', ZOOM), ('PDF', PDF), ('Combined', COMBINED)]:
    count = search_client.aggregate_data_objects(
        vectorsearch_v1beta.AggregateDataObjectsRequest(parent=parent, aggregate='COUNT')
    )
    print(f'{label}: {int(count.aggregate_results[0]["count"])} DataObjects')

Reddit: 34 DataObjects
Zoom: 12 DataObjects
PDF: 71 DataObjects
Combined: 117 DataObjects


---
# Query

**Query** filters DataObjects by their data fields — analogous to SQL `WHERE`. No relevance ranking — results are unordered. Supports comparison operators (`$eq`, `$gt`, `$lt`, `$gte`, `$lte`), logical operators (`$and`, `$or`), and array operators (`$in`, `$nin`).

In [5]:
# Filter by source_type
results = list(search_client.query_data_objects(
    vectorsearch_v1beta.QueryDataObjectsRequest(
        parent=COMBINED,
        filter={'source_type': {'$eq': 'zoom'}},
        output_fields=vectorsearch_v1beta.OutputFields(
            data_fields=['chunk_id', 'source_type', 'source_uri', 'text_content']
        ),
    )
))
print(f'{len(results)} zoom chunks')
query_df(results[:5], ['chunk_id', 'source_type', 'source_uri', 'text_content'])

12 zoom chunks


,chunk_id,source_type,source_uri,text_content
0,zoom_meeting_model_planning_001,zoom,gs://statmike-mlops-349915/applied-genai-solut...,[Summary: The team debates Prophet versus a gr...
1,zoom_meeting_model_planning_003,zoom,gs://statmike-mlops-349915/applied-genai-solut...,[Summary: The team finalized a plan to benchma...
2,zoom_meeting_business_readout_000,zoom,gs://statmike-mlops-349915/applied-genai-solut...,"Tom Davis: Alright everyone, thanks for joinin..."
3,zoom_meeting_model_planning_000,zoom,gs://statmike-mlops-349915/applied-genai-solut...,"Sarah (Lead): Alright everyone, thanks for joi..."
4,zoom_meeting_business_readout_003,zoom,gs://statmike-mlops-349915/applied-genai-solut...,[Summary: The team agreed to provide cleaner d...


In [6]:
# Filter by karma threshold (Reddit-specific)
results = list(search_client.query_data_objects(
    vectorsearch_v1beta.QueryDataObjectsRequest(
        parent=COMBINED,
        filter={'karma': {'$gt': 50}},
        output_fields=vectorsearch_v1beta.OutputFields(
            data_fields=['chunk_id', 'source_type', 'karma', 'text_content']
        ),
    )
))
print(f'{len(results)} chunks with karma > 50')
query_df(results, ['chunk_id', 'source_type', 'karma', 'text_content'])

7 chunks with karma > 50


,chunk_id,source_type,karma,text_content
0,red_thread_shallow_003,reddit,52.0,THREAD: ARIMA vs. Prophet for Retail Demand Fo...
1,red_thread_medium_000,reddit,98.0,THREAD: Beyond ARIMA and Prophet: What are you...
2,red_thread_shallow_000,reddit,89.0,THREAD: ARIMA vs. Prophet for Retail Demand Fo...
3,red_thread_deep_000,reddit,82.0,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
4,red_thread_deep_007,reddit,53.0,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
5,red_thread_deep_001,reddit,121.0,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
6,red_thread_medium_004,reddit,52.0,THREAD: Beyond ARIMA and Prophet: What are you...


In [7]:
# Filter by page range (PDF-specific)
results = list(search_client.query_data_objects(
    vectorsearch_v1beta.QueryDataObjectsRequest(
        parent=COMBINED,
        filter={'page_start': {'$lte': 2}},
        output_fields=vectorsearch_v1beta.OutputFields(
            data_fields=['chunk_id', 'source_type', 'page_start', 'page_end', 'text_content']
        ),
    )
))
print(f'{len(results)} chunks from pages 0-2')
query_df(results[:5], ['chunk_id', 'source_type', 'page_start', 'page_end', 'text_content'])

18 chunks from pages 0-2


,chunk_id,source_type,page_start,page_end,text_content
0,pdf_944a4f7390_c3,pdf,1.0,4.0,# Automatic feature preprocessing\n\n## Missin...
1,pdf_0712f4777d_c2,pdf,2.0,5.0,# The ML.EXPLAIN_FORECAST function\n\n## Outpu...
2,pdf_5e3c207262_c1,pdf,1.0,3.0,# The ML.HOLIDAY INFO function\n\nThis documen...
3,pdf_765aeabdb0_c1,pdf,1.0,3.0,# The ML.ARIMA_COEFFICIENTS function\n\nThis d...
4,pdf_90bbc35dad_c2,pdf,1.0,3.0,# The CREATE MODEL statement for ARIMA PLUS mo...


In [8]:
# Filter by enrichment topic ($in operator on array field)
results = list(search_client.query_data_objects(
    vectorsearch_v1beta.QueryDataObjectsRequest(
        parent=COMBINED,
        filter={'topics': {'$in': ['time series forecasting']}},
        output_fields=vectorsearch_v1beta.OutputFields(
            data_fields=['chunk_id', 'source_type', 'topics', 'text_content']
        ),
    )
))
print(f'{len(results)} chunks tagged with "time series forecasting"')

from collections import Counter as _Counter
type_counts = _Counter(r.data['source_type'] for r in results)
for stype, count in type_counts.items():
    print(f'  {stype}: {count}')

query_df(results[:5], ['chunk_id', 'source_type', 'topics', 'text_content'])

24 chunks tagged with "time series forecasting"
  pdf: 15
  reddit: 9


,chunk_id,source_type,topics,text_content
0,pdf_5e3c207262_c1,pdf,"ARIMA_PLUS, ARIMA_PLUS_XREG, time series forec...",# The ML.HOLIDAY INFO function\n\nThis documen...
1,pdf_90bbc35dad_c2,pdf,"ARIMA PLUS, time series modeling pipeline, Big...",# The CREATE MODEL statement for ARIMA PLUS mo...
2,pdf_5fa7c6b197_c3,pdf,"ARIMA PLUS XREG models, time series forecastin...",# The CREATE MODEL statement for ARIMA PLUS XR...
3,red_thread_shallow_005,reddit,"demand forecasting, business forecasting, expl...",THREAD: ARIMA vs. Prophet for Retail Demand Fo...
4,red_thread_medium_000,reddit,"time series forecasting, complex data, seasona...",THREAD: Beyond ARIMA and Prophet: What are you...


---
# Single-Channel Search

Each of these uses `search_data_objects` with a single search channel.

### SearchHint (ANN vs kNN)

`VectorSearch` and `SemanticSearch` accept a `search_hint` parameter to control index usage. By default, VS2 auto-selects an ANN index if one exists, falling back to brute-force kNN. Use `SearchHint(use_knn=True)` to force exact kNN search (useful for recall evaluation). `TextSearch` does not support `search_hint`.

| Behavior | How to set |
|----------|-----------|
| **Auto (default)** | Omit `search_hint` — uses ANN if available, falls back to kNN |
| **Force kNN** | `search_hint=SearchHint(use_knn=True)` |
| **Force specific index** | `search_hint=SearchHint(use_index=IndexHint(name=...))` |

`SearchResponseMetadata` in the response reports which engine was used. See step 10 (`manage_ann_index.py` and `inspect_ann_index.ipynb`) for ANN index management and recall benchmarking.

## SemanticSearch

You provide query text, VS2 auto-embeds it using the model configured in the collection's vector schema. The simplest search — no manual embedding needed.

Notice the results below are all Reddit chunks — semantic search surfaces community discussions that match the *meaning* of "demand forecasting" even though the PDF documentation also covers this topic. This is because semantic similarity captures conversational relevance, which community threads express strongly.

In [9]:
query = 'What machine learning methods work best for demand forecasting?'

results = search_client.search_data_objects(
    vectorsearch_v1beta.SearchDataObjectsRequest(
        parent=COMBINED,
        semantic_search=vectorsearch_v1beta.SemanticSearch(
            search_text=query,
            search_field='text_content_embedding',
            task_type='QUESTION_ANSWERING',
            top_k=5,
            output_fields=vectorsearch_v1beta.OutputFields(
                data_fields=['chunk_id', 'source_type', 'text_content']
            ),
            # To force kNN (brute-force) instead of auto-selecting ANN:
            #   search_hint=vectorsearch_v1beta.SearchHint(use_knn=True)
            # To force a specific ANN index:
            #   search_hint=vectorsearch_v1beta.SearchHint(
            #       use_index=vectorsearch_v1beta.SearchHint.IndexHint(
            #           name=f"{COMBINED}/indexes/{ANN_INDEX_ID}"
            #       )
            #   )
        ),
    )
)

print(f'SemanticSearch: "{query}"\n')
search_df(results, ['chunk_id', 'source_type', 'text_content'])

SemanticSearch: "What machine learning methods work best for demand forecasting?"



,chunk_id,source_type,text_content
rank,,,
1,red_thread_deep_015,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
2,red_thread_shallow_003,reddit,THREAD: ARIMA vs. Prophet for Retail Demand Fo...
3,red_thread_medium_000,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
4,red_thread_deep_016,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
5,red_thread_medium_001,reddit,THREAD: Beyond ARIMA and Prophet: What are you...


## TextSearch

Native keyword matching. Useful for specific method names, acronyms, or terms that may not have strong semantic representation.

Compare to SemanticSearch above: searching "ARIMA" as a keyword surfaces the PDF technical documentation (where the term appears frequently and prominently), while semantic search for a question about "demand forecasting" favored the Reddit discussions. This illustrates why hybrid search is valuable — each channel finds different relevant content.

In [10]:
query = 'ARIMA'

results = search_client.search_data_objects(
    vectorsearch_v1beta.SearchDataObjectsRequest(
        parent=COMBINED,
        text_search=vectorsearch_v1beta.TextSearch(
            search_text=query,
            data_field_names=['text_content'],
            top_k=5,
            output_fields=vectorsearch_v1beta.OutputFields(
                data_fields=['chunk_id', 'source_type', 'text_content']
            ),
        ),
    )
)

print(f'TextSearch: "{query}"\n')
search_df(results, ['chunk_id', 'source_type', 'text_content'])

TextSearch: "ARIMA"



,chunk_id,source_type,text_content
rank,,,
1,pdf_90bbc35dad_c2,pdf,# The CREATE MODEL statement for ARIMA PLUS mo...
2,pdf_90bbc35dad_c5,pdf,# model_name\n\nThe name of the model you're c...
3,pdf_90bbc35dad_c19,pdf,# What's next\n\n• Try the following tutorials...
4,pdf_5fa7c6b197_c1,pdf,# The CREATE MODEL statement for ARIMA PLUS XR...
5,pdf_2f59757db2_c2,pdf,# The ML.ARIMA EVALUATE function\n\n## Output\...


## VectorSearch (dense only)

Supply a dense embedding vector manually. This produces nearly identical results to SemanticSearch when using the same model and task type — validating that auto-embedding and manual embedding are equivalent. The same set of chunks appears in the top-5, though ranking order may differ slightly due to internal precision differences. Required when using embeddings from external models.

**Important:** You must set `output_dimensionality` to match the collection schema (768 in this project). `gemini-embedding-001` natively produces 3072-dim vectors, but VS2 auto-embeds at the schema dimension internally.

In [11]:
query = 'What machine learning methods work best for demand forecasting?'

# Generate embedding with output_dimensionality matching the collection schema (768).
# gemini-embedding-001 natively produces 3072-dim vectors; VS2 auto-embeds at the schema
# dimension, so manual embeddings must request the same size.
embedding_response = genai_client.models.embed_content(
    model=EMBEDDING_MODEL,
    contents=query,
    config={'task_type': 'QUESTION_ANSWERING', 'output_dimensionality': EMBEDDING_DIMENSIONS},
)
dense_vector = embedding_response.embeddings[0].values

results = search_client.search_data_objects(
    vectorsearch_v1beta.SearchDataObjectsRequest(
        parent=COMBINED,
        vector_search=vectorsearch_v1beta.VectorSearch(
            search_field='text_content_embedding',
            vector=vectorsearch_v1beta.DenseVector(values=dense_vector),
            top_k=5,
            output_fields=vectorsearch_v1beta.OutputFields(
                data_fields=['chunk_id', 'source_type', 'text_content']
            ),
        ),
    )
)

print(f'VectorSearch (dense): "{query}"')
print(f'Dense vector: {len(dense_vector)} dimensions\n')
search_df(results, ['chunk_id', 'source_type', 'text_content'])

VectorSearch (dense): "What machine learning methods work best for demand forecasting?"
Dense vector: 768 dimensions



,chunk_id,source_type,text_content
rank,,,
1,red_thread_deep_015,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
2,red_thread_medium_000,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
3,red_thread_shallow_003,reddit,THREAD: ARIMA vs. Prophet for Retail Demand Fo...
4,red_thread_medium_001,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
5,red_thread_deep_016,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...


## VectorSearch (sparse only)

Supply a BM25 sparse embedding. This uses the vocabulary trained in step 7a (`build_bm25.py`) and applied in step 7b (`apply_bm25.py`).

Compare to TextSearch above — both do keyword-style matching but use different scoring. BM25 weights terms by IDF (rarer terms score higher) and normalizes for document length, while TextSearch uses VS2's native keyword matching. Notice the results overlap but rank differently: BM25 may surface chunks where "ARIMA" appears in a more discriminating context.

In [12]:
query = 'ARIMA time series model training'

indices, values = create_bm25_sparse_embedding(query)

results = search_client.search_data_objects(
    vectorsearch_v1beta.SearchDataObjectsRequest(
        parent=COMBINED,
        vector_search=vectorsearch_v1beta.VectorSearch(
            search_field='bm25_embedding',
            sparse_vector=vectorsearch_v1beta.SparseVector(values=values, indices=indices),
            top_k=5,
            output_fields=vectorsearch_v1beta.OutputFields(
                data_fields=['chunk_id', 'source_type', 'text_content']
            ),
        ),
    )
)

print(f'VectorSearch (sparse/BM25): "{query}"')
print(f'Sparse vector: {len(indices)} non-zero terms\n')
search_df(results, ['chunk_id', 'source_type', 'text_content'])

VectorSearch (sparse/BM25): "ARIMA time series model training"
Sparse vector: 8 non-zero terms



,chunk_id,source_type,text_content
rank,,,
1,pdf_90bbc35dad_c18,pdf,# Limitations\n\nARIMA_PLUS models have the fo...
2,pdf_5fa7c6b197_c3,pdf,# The CREATE MODEL statement for ARIMA PLUS XR...
3,pdf_90bbc35dad_c3,pdf,# The CREATE MODEL statement for ARIMA PLUS mo...
4,pdf_1affd50072_c1,pdf,# The ML.DETECT_ANOMALIES function\n\nThis doc...
5,pdf_0712f4777d_c1,pdf,# The ML.EXPLAIN_FORECAST function\n\nThis doc...


## VectorSearch (dense + sparse via RRF)

The `VectorSearch` proto uses a `oneof` for `vector` and `sparse_vector` — you can supply one or the other per search channel, not both. To combine dense and sparse vectors, use `batch_search_data_objects` with two VectorSearch channels fused via RRF. This gives explicit weight control over how much each vector type contributes.

Notice the results below mix PDF chunks (from sparse/BM25, which favors technical docs with exact terms) and Reddit chunks (from dense/semantic, which favors discussion threads). This is the value of combining both vector types — each surfaces different relevant content.

In [13]:
query = 'ARIMA time series model training'

# Dense embedding (must match collection schema dimensions)
embedding_response = genai_client.models.embed_content(
    model=EMBEDDING_MODEL,
    contents=query,
    config={'task_type': 'QUESTION_ANSWERING', 'output_dimensionality': EMBEDDING_DIMENSIONS},
)
dense_vector = embedding_response.embeddings[0].values

# Sparse embedding
sparse_indices, sparse_values = create_bm25_sparse_embedding(query)

# Use batch_search with two VectorSearch channels fused via RRF
# (VectorSearch uses a oneof — vector and sparse_vector are mutually exclusive per channel)
fields = ['chunk_id', 'source_type', 'text_content']
output = vectorsearch_v1beta.OutputFields(data_fields=fields)

response = search_client.batch_search_data_objects(
    vectorsearch_v1beta.BatchSearchDataObjectsRequest(
        parent=COMBINED,
        searches=[
            vectorsearch_v1beta.Search(
                vector_search=vectorsearch_v1beta.VectorSearch(
                    search_field='text_content_embedding',
                    vector=vectorsearch_v1beta.DenseVector(values=dense_vector),
                    top_k=5,
                    output_fields=output,
                )
            ),
            vectorsearch_v1beta.Search(
                vector_search=vectorsearch_v1beta.VectorSearch(
                    search_field='bm25_embedding',
                    sparse_vector=vectorsearch_v1beta.SparseVector(values=sparse_values, indices=sparse_indices),
                    top_k=5,
                    output_fields=output,
                )
            ),
        ],
        combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
            ranker=vectorsearch_v1beta.Ranker(
                rrf=vectorsearch_v1beta.ReciprocalRankFusion(weights=[1.0, 1.0])
            )
        ),
    )
)

results = list(response.results[0].results) if response.results else []
print(f'VectorSearch (dense + sparse via RRF): "{query}"')
print(f'Dense: {len(dense_vector)} dims, Sparse: {len(sparse_indices)} terms\n')
search_df(results[:5], fields)

VectorSearch (dense + sparse via RRF): "ARIMA time series model training"
Dense: 768 dims, Sparse: 8 terms



,chunk_id,source_type,text_content
rank,,,
1,pdf_90bbc35dad_c3,pdf,# The CREATE MODEL statement for ARIMA PLUS mo...
2,pdf_5fa7c6b197_c3,pdf,# The CREATE MODEL statement for ARIMA PLUS XR...
3,red_thread_deep_000,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
4,pdf_90bbc35dad_c18,pdf,# Limitations\n\nARIMA_PLUS models have the fo...
5,red_thread_shallow_000,reddit,THREAD: ARIMA vs. Prophet for Retail Demand Fo...


## Filtered Search

The `filter` parameter on `SemanticSearch` applies a pre-filter before vector similarity ranking. This restricts results to a metadata subset without affecting embedding comparison quality.

Here we filter to PDF-only, which addresses the source-type bias seen in unfiltered SemanticSearch above. An alternative to filtering is the crowding pattern shown later, which retrieves top results from each group in parallel.

In [14]:
query = 'What methods handle seasonal patterns in demand forecasting?'

# Same query, filtered to PDF-only results
results = search_client.search_data_objects(
    vectorsearch_v1beta.SearchDataObjectsRequest(
        parent=COMBINED,
        semantic_search=vectorsearch_v1beta.SemanticSearch(
            search_text=query,
            search_field='text_content_embedding',
            task_type='QUESTION_ANSWERING',
            top_k=5,
            output_fields=vectorsearch_v1beta.OutputFields(
                data_fields=['chunk_id', 'source_type', 'source_uri', 'text_content']
            ),
            filter={'source_type': {'$eq': 'pdf'}},
        ),
    )
)

print(f'Filtered SemanticSearch (pdf only): "{query}"\n')
search_df(results, ['chunk_id', 'source_type', 'source_uri', 'text_content'])

Filtered SemanticSearch (pdf only): "What methods handle seasonal patterns in demand forecasting?"



,chunk_id,source_type,source_uri,text_content
rank,,,,
1,pdf_90bbc35dad_c19,pdf,gs://statmike-mlops-349915/applied-genai-solut...,# What's next\n\n• Try the following tutorials...
2,pdf_90bbc35dad_c16,pdf,gs://statmike-mlops-349915/applied-genai-solut...,# Custom holiday limitations\n\n• Custom holid...
3,pdf_90bbc35dad_c17,pdf,gs://statmike-mlops-349915/applied-genai-solut...,# Hierarchical reconciliation\n\nYou can break...
4,pdf_1fa007c733_c4,pdf,gs://statmike-mlops-349915/applied-genai-solut...,# The Al.FORECAST function\n\n## Examples\n\nT...
5,pdf_5fa7c6b197_c13,pdf,gs://statmike-mlops-349915/applied-genai-solut...,# Holiday data\n\n|-|-|-|-|-|\n| region | holi...


---
# Hybrid Search (RRF)

Hybrid search runs multiple search channels in parallel, then combines results using [Reciprocal Rank Fusion (RRF)](https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf). Each result's final score is:

```
score(d) = SUM( weight_i / (k + rank_i(d)) )
```

where `k=60` (standard), `rank_i(d)` is the document's rank in channel `i`, and `weight_i` scales that channel's contribution. RRF is purely statistical — it merges ranked lists without looking at the actual scores, making it robust to score distribution differences across channels.

## Semantic + Text (2-way)

The classic hybrid pattern: semantic understanding + keyword matching. Uses the `hybrid_search` helper function defined in Setup.

In [15]:
query = 'seasonal ARIMA forecasting methods'
fields = ['chunk_id', 'source_type', 'text_content']

results = hybrid_search(COMBINED, query, weights=[1.0, 1.0], top_k=5, fields=fields)

print(f'Hybrid Semantic+Text [1, 1]: "{query}"\n')
search_df(results, fields)

Hybrid Semantic+Text [1, 1]: "seasonal ARIMA forecasting methods"



,chunk_id,source_type,text_content
rank,,,
1,zoom_meeting_model_planning_000,zoom,"Sarah (Lead): Alright everyone, thanks for joi..."
2,red_thread_deep_000,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
3,red_thread_medium_004,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
4,red_thread_medium_000,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
5,red_thread_deep_003,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
6,red_thread_deep_004,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
7,red_thread_medium_007,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
8,red_thread_deep_011,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
9,zoom_meeting_model_planning_001,zoom,[Summary: The team debates Prophet versus a gr...


## Semantic + Sparse (2-way)

Replace the TextSearch channel with a VectorSearch sparse (BM25) channel. BM25 captures term importance via IDF and document length normalization, while TextSearch uses VS2's native keyword matching.

Compare rankings to Semantic+Text above — the same query produces different top results because BM25 scores terms differently than native text search. BM25's enrichment tags (topics, methods) also contribute to matching, which can surface chunks that wouldn't rank highly on raw text alone.

In [16]:
query = 'seasonal ARIMA forecasting methods'
fields = ['chunk_id', 'source_type', 'text_content']

indices, values = create_bm25_sparse_embedding(query)
output = vectorsearch_v1beta.OutputFields(data_fields=fields)

response = search_client.batch_search_data_objects(
    vectorsearch_v1beta.BatchSearchDataObjectsRequest(
        parent=COMBINED,
        searches=[
            vectorsearch_v1beta.Search(
                semantic_search=vectorsearch_v1beta.SemanticSearch(
                    search_text=query,
                    search_field='text_content_embedding',
                    task_type='QUESTION_ANSWERING',
                    top_k=10,
                    output_fields=output,
                )
            ),
            vectorsearch_v1beta.Search(
                vector_search=vectorsearch_v1beta.VectorSearch(
                    search_field='bm25_embedding',
                    sparse_vector=vectorsearch_v1beta.SparseVector(values=values, indices=indices),
                    top_k=10,
                    output_fields=output,
                )
            ),
        ],
        combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
            ranker=vectorsearch_v1beta.Ranker(
                rrf=vectorsearch_v1beta.ReciprocalRankFusion(weights=[1.0, 1.0])
            )
        ),
    )
)

results = list(response.results[0].results) if response.results else []
print(f'Hybrid Semantic+Sparse [1, 1]: "{query}"\n')
search_df(results[:5], fields)

Hybrid Semantic+Sparse [1, 1]: "seasonal ARIMA forecasting methods"



,chunk_id,source_type,text_content
rank,,,
1,red_thread_medium_004,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
2,red_thread_medium_000,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
3,zoom_meeting_model_planning_000,zoom,"Sarah (Lead): Alright everyone, thanks for joi..."
4,red_thread_deep_000,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
5,red_thread_deep_003,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...


## 3-Way: Semantic + Text + Sparse

All three channels combined with RRF. Each captures a different signal:
- **Semantic**: Meaning and intent (finds synonyms, related concepts)
- **Text**: Exact keyword matching (VS2 native)
- **Sparse (BM25)**: Statistical term importance with IDF weighting and document length normalization

In [17]:
query = 'seasonal ARIMA forecasting methods'
fields = ['chunk_id', 'source_type', 'text_content']

indices, values = create_bm25_sparse_embedding(query)
output = vectorsearch_v1beta.OutputFields(data_fields=fields)

response = search_client.batch_search_data_objects(
    vectorsearch_v1beta.BatchSearchDataObjectsRequest(
        parent=COMBINED,
        searches=[
            vectorsearch_v1beta.Search(
                semantic_search=vectorsearch_v1beta.SemanticSearch(
                    search_text=query,
                    search_field='text_content_embedding',
                    task_type='QUESTION_ANSWERING',
                    top_k=10,
                    output_fields=output,
                )
            ),
            vectorsearch_v1beta.Search(
                text_search=vectorsearch_v1beta.TextSearch(
                    search_text=query,
                    data_field_names=['text_content'],
                    top_k=10,
                    output_fields=output,
                )
            ),
            vectorsearch_v1beta.Search(
                vector_search=vectorsearch_v1beta.VectorSearch(
                    search_field='bm25_embedding',
                    sparse_vector=vectorsearch_v1beta.SparseVector(values=values, indices=indices),
                    top_k=10,
                    output_fields=output,
                )
            ),
        ],
        combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
            ranker=vectorsearch_v1beta.Ranker(
                rrf=vectorsearch_v1beta.ReciprocalRankFusion(weights=[1.0, 1.0, 1.0])
            )
        ),
    )
)

results = list(response.results[0].results) if response.results else []
print(f'Hybrid 3-way Semantic+Text+Sparse [1, 1, 1]: "{query}"\n')
search_df(results[:8], fields)

Hybrid 3-way Semantic+Text+Sparse [1, 1, 1]: "seasonal ARIMA forecasting methods"



,chunk_id,source_type,text_content
rank,,,
1,red_thread_medium_004,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
2,red_thread_medium_000,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
3,zoom_meeting_model_planning_000,zoom,"Sarah (Lead): Alright everyone, thanks for joi..."
4,red_thread_medium_007,reddit,THREAD: Beyond ARIMA and Prophet: What are you...
5,red_thread_deep_000,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
6,red_thread_deep_003,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
7,red_thread_deep_004,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
8,red_thread_medium_003,reddit,THREAD: Beyond ARIMA and Prophet: What are you...


## RRF Weight Tuning

Compare how different weight configurations shift the ranking. Lower rank = higher relevance.

The 2-way table below shows that weighting Text `[1, 3]` pulls the Zoom meeting chunk higher (it contains many keyword matches), while weighting Semantic `[3, 1]` keeps it near the top but lets conversational Reddit threads compete more effectively. The 3-way table adds BM25 sparse as a third signal — the Keyword-heavy config `[1, 2, 2]` shifts rankings significantly because both TextSearch and BM25 now reinforce keyword-rich chunks.

In [18]:
query = 'seasonal ARIMA forecasting methods'
fields = ['chunk_id', 'text_content']

# 2-way Semantic+Text weight comparison
configs = {
    'Equal [1, 1]': [1.0, 1.0],
    'Semantic [3, 1]': [3.0, 1.0],
    'Text [1, 3]': [1.0, 3.0],
}

rankings = {}
for label, weights in configs.items():
    results = hybrid_search(COMBINED, query, weights, top_k=10, fields=fields)
    for rank, r in enumerate(results, 1):
        cid = r.data_object.data['chunk_id']
        if cid not in rankings:
            text = r.data_object.data['text_content']
            rankings[cid] = {'chunk_id': cid, 'text_preview': text[:100] + '...'}
        rankings[cid][label] = rank

comparison = pd.DataFrame(rankings.values()).sort_values('Equal [1, 1]')
print(f'RRF weight comparison (Semantic+Text): "{query}"')
print('Lower rank = higher relevance.\n')
comparison

RRF weight comparison (Semantic+Text): "seasonal ARIMA forecasting methods"
Lower rank = higher relevance.



,chunk_id,text_preview,"Equal [1, 1]","Semantic [3, 1]","Text [1, 3]"
0,red_thread_medium_004,THREAD: Beyond ARIMA and Prophet: What are you...,1,2,1
1,zoom_meeting_model_planning_000,"Sarah (Lead): Alright everyone, thanks for joi...",2,1,3
2,red_thread_medium_000,THREAD: Beyond ARIMA and Prophet: What are you...,3,3,2
3,red_thread_deep_000,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,4,4,6
4,red_thread_deep_003,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,5,5,7
5,red_thread_deep_004,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,6,6,8
6,red_thread_medium_007,THREAD: Beyond ARIMA and Prophet: What are you...,7,11,4
7,red_thread_deep_011,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,8,7,9
8,zoom_meeting_model_planning_001,[Summary: The team debates Prophet versus a gr...,9,12,5
9,red_thread_shallow_000,THREAD: ARIMA vs. Prophet for Retail Demand Fo...,10,8,10


In [19]:
# 3-way weight comparison
query = 'seasonal ARIMA forecasting methods'
fields = ['chunk_id', 'text_content']
output = vectorsearch_v1beta.OutputFields(data_fields=fields)

indices, values = create_bm25_sparse_embedding(query)

configs_3way = {
    'Equal [1,1,1]': [1.0, 1.0, 1.0],
    'Sem-heavy [3,1,1]': [3.0, 1.0, 1.0],
    'Keyword-heavy [1,2,2]': [1.0, 2.0, 2.0],
}

rankings_3 = {}
for label, weights in configs_3way.items():
    response = search_client.batch_search_data_objects(
        vectorsearch_v1beta.BatchSearchDataObjectsRequest(
            parent=COMBINED,
            searches=[
                vectorsearch_v1beta.Search(
                    semantic_search=vectorsearch_v1beta.SemanticSearch(
                        search_text=query, search_field='text_content_embedding',
                        task_type='QUESTION_ANSWERING', top_k=10, output_fields=output,
                    )
                ),
                vectorsearch_v1beta.Search(
                    text_search=vectorsearch_v1beta.TextSearch(
                        search_text=query, data_field_names=['text_content'],
                        top_k=10, output_fields=output,
                    )
                ),
                vectorsearch_v1beta.Search(
                    vector_search=vectorsearch_v1beta.VectorSearch(
                        search_field='bm25_embedding',
                        sparse_vector=vectorsearch_v1beta.SparseVector(values=values, indices=indices),
                        top_k=10, output_fields=output,
                    )
                ),
            ],
            combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
                ranker=vectorsearch_v1beta.Ranker(
                    rrf=vectorsearch_v1beta.ReciprocalRankFusion(weights=weights)
                )
            ),
        )
    )
    results = list(response.results[0].results) if response.results else []
    for rank, r in enumerate(results, 1):
        cid = r.data_object.data['chunk_id']
        if cid not in rankings_3:
            text = r.data_object.data['text_content']
            rankings_3[cid] = {'chunk_id': cid, 'text_preview': text[:80] + '...'}
        rankings_3[cid][label] = rank

comparison_3 = pd.DataFrame(rankings_3.values()).sort_values('Equal [1,1,1]')
print(f'RRF weight comparison (3-way Semantic+Text+Sparse): "{query}"')
print('Lower rank = higher relevance.\n')
comparison_3

RRF weight comparison (3-way Semantic+Text+Sparse): "seasonal ARIMA forecasting methods"
Lower rank = higher relevance.



,chunk_id,text_preview,"Equal [1,1,1]","Sem-heavy [3,1,1]","Keyword-heavy [1,2,2]"
0,red_thread_medium_004,THREAD: Beyond ARIMA and Prophet: What are you...,1,1,1
1,red_thread_medium_000,THREAD: Beyond ARIMA and Prophet: What are you...,2,3,2
2,zoom_meeting_model_planning_000,"Sarah (Lead): Alright everyone, thanks for joi...",3,2,3
3,red_thread_medium_007,THREAD: Beyond ARIMA and Prophet: What are you...,4,11,4
4,red_thread_deep_000,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,5,4,12
5,red_thread_deep_003,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,6,5,13
6,red_thread_deep_004,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,7,6,14
7,red_thread_medium_003,THREAD: Beyond ARIMA and Prophet: What are you...,8,12,5
8,red_thread_deep_011,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,9,7,15
9,red_thread_medium_008,THREAD: Beyond ARIMA and Prophet: What are you...,10,13,6


---
# Reranking with VertexRanker

> **Note:** VertexRanker is defined in the VS2 API proto (`vectorsearch_v1beta.VertexRanker`) but **not yet supported** by the backend — requests currently return `UNIMPLEMENTED: Only RRF is supported`. The cells below show the intended API pattern for when it becomes available. They catch the error gracefully and print the status.

VertexRanker is a model-based cross-encoder that scores each (query, document) pair — more accurate but slower than RRF, which is purely statistical. Use `batch_search_data_objects` with `VertexRanker` as the ranker instead of RRF.

## Single Channel + VertexRanker

A single SemanticSearch channel reranked by VertexRanker. The initial retrieval uses embedding similarity, then VertexRanker rescores the top results using a cross-encoder.

In [20]:
query = 'What machine learning methods work best for demand forecasting?'
fields = ['chunk_id', 'source_type', 'text_content']
output = vectorsearch_v1beta.OutputFields(data_fields=fields)

try:
    response = search_client.batch_search_data_objects(
        vectorsearch_v1beta.BatchSearchDataObjectsRequest(
            parent=COMBINED,
            searches=[
                vectorsearch_v1beta.Search(
                    semantic_search=vectorsearch_v1beta.SemanticSearch(
                        search_text=query,
                        search_field='text_content_embedding',
                        task_type='QUESTION_ANSWERING',
                        top_k=10,
                        output_fields=output,
                    )
                ),
            ],
            combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
                ranker=vectorsearch_v1beta.Ranker(
                    vertex=vectorsearch_v1beta.VertexRanker(
                        query=query,
                        content_template='{text_content}',
                    )
                )
            ),
        )
    )
    results = list(response.results[0].results) if response.results else []
    print(f'Semantic + VertexRanker: "{query}"\n')
    search_df(results[:5], fields)
except Exception as e:
    print(f'VertexRanker not yet supported: {e.message if hasattr(e, "message") else e}')

VertexRanker not yet supported: Operation is not implemented, or supported, or enabled.


## Hybrid + VertexRanker

Multiple search channels reranked by VertexRanker instead of RRF. The cross-encoder rescores candidates from all channels based on actual query-document relevance.

In [21]:
query = 'What machine learning methods work best for demand forecasting?'
fields = ['chunk_id', 'source_type', 'text_content']
output = vectorsearch_v1beta.OutputFields(data_fields=fields)

try:
    response = search_client.batch_search_data_objects(
        vectorsearch_v1beta.BatchSearchDataObjectsRequest(
            parent=COMBINED,
            searches=[
                vectorsearch_v1beta.Search(
                    semantic_search=vectorsearch_v1beta.SemanticSearch(
                        search_text=query,
                        search_field='text_content_embedding',
                        task_type='QUESTION_ANSWERING',
                        top_k=10,
                        output_fields=output,
                    )
                ),
                vectorsearch_v1beta.Search(
                    text_search=vectorsearch_v1beta.TextSearch(
                        search_text=query,
                        data_field_names=['text_content'],
                        top_k=10,
                        output_fields=output,
                    )
                ),
            ],
            combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
                ranker=vectorsearch_v1beta.Ranker(
                    vertex=vectorsearch_v1beta.VertexRanker(
                        query=query,
                        content_template='{text_content}',
                    )
                )
            ),
        )
    )
    results = list(response.results[0].results) if response.results else []
    print(f'Hybrid (Semantic+Text) + VertexRanker: "{query}"\n')
    search_df(results[:5], fields)
except Exception as e:
    print(f'VertexRanker not yet supported: {e.message if hasattr(e, "message") else e}')

VertexRanker not yet supported: Operation is not implemented, or supported, or enabled.


## RRF vs VertexRanker Comparison

Same query, same channels — compare how RRF (statistical rank fusion) and VertexRanker (model-based cross-encoder) produce different final rankings. Since VertexRanker is not yet supported, the comparison below shows RRF results only. When VertexRanker becomes available, a VertexRanker column will appear alongside RRF for side-by-side comparison.

In [22]:
query = 'What machine learning methods work best for demand forecasting?'
fields = ['chunk_id', 'text_content']
output = vectorsearch_v1beta.OutputFields(data_fields=fields)

searches = [
    vectorsearch_v1beta.Search(
        semantic_search=vectorsearch_v1beta.SemanticSearch(
            search_text=query,
            search_field='text_content_embedding',
            task_type='QUESTION_ANSWERING',
            top_k=10,
            output_fields=output,
        )
    ),
    vectorsearch_v1beta.Search(
        text_search=vectorsearch_v1beta.TextSearch(
            search_text=query,
            data_field_names=['text_content'],
            top_k=10,
            output_fields=output,
        )
    ),
]

# RRF
rrf_response = search_client.batch_search_data_objects(
    vectorsearch_v1beta.BatchSearchDataObjectsRequest(
        parent=COMBINED, searches=searches,
        combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
            ranker=vectorsearch_v1beta.Ranker(
                rrf=vectorsearch_v1beta.ReciprocalRankFusion(weights=[1.0, 1.0])
            )
        ),
    )
)

rrf_results = list(rrf_response.results[0].results) if rrf_response.results else []

# VertexRanker
try:
    vr_response = search_client.batch_search_data_objects(
        vectorsearch_v1beta.BatchSearchDataObjectsRequest(
            parent=COMBINED, searches=searches,
            combine=vectorsearch_v1beta.BatchSearchDataObjectsRequest.CombineResultsOptions(
                ranker=vectorsearch_v1beta.Ranker(
                    vertex=vectorsearch_v1beta.VertexRanker(
                        query=query, content_template='{text_content}',
                    )
                )
            ),
        )
    )
    vr_results = list(vr_response.results[0].results) if vr_response.results else []
except Exception as e:
    print(f'VertexRanker not yet supported: {e.message if hasattr(e, "message") else e}')
    vr_results = []

rankings = {}
for rank, r in enumerate(rrf_results, 1):
    cid = r.data_object.data['chunk_id']
    text = r.data_object.data['text_content']
    rankings[cid] = {'chunk_id': cid, 'text_preview': text[:80] + '...', 'RRF [1,1]': rank}
for rank, r in enumerate(vr_results, 1):
    cid = r.data_object.data['chunk_id']
    if cid not in rankings:
        text = r.data_object.data['text_content']
        rankings[cid] = {'chunk_id': cid, 'text_preview': text[:80] + '...'}
    rankings[cid]['VertexRanker'] = rank

comparison = pd.DataFrame(rankings.values()).sort_values('RRF [1,1]')
print(f'RRF vs VertexRanker: "{query}"')
if vr_results:
    print('Lower rank = higher relevance.\n')
else:
    print('VertexRanker column omitted (not yet supported). Showing RRF results only.\n')
comparison

VertexRanker not yet supported: Operation is not implemented, or supported, or enabled.
RRF vs VertexRanker: "What machine learning methods work best for demand forecasting?"
VertexRanker column omitted (not yet supported). Showing RRF results only.



,chunk_id,text_preview,"RRF [1,1]"
0,red_thread_deep_015,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,1
1,red_thread_shallow_003,THREAD: ARIMA vs. Prophet for Retail Demand Fo...,2
2,red_thread_medium_000,THREAD: Beyond ARIMA and Prophet: What are you...,3
3,red_thread_deep_016,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,4
4,red_thread_medium_001,THREAD: Beyond ARIMA and Prophet: What are you...,5
5,zoom_meeting_model_planning_000,"Sarah (Lead): Alright everyone, thanks for joi...",6
6,red_thread_shallow_004,THREAD: ARIMA vs. Prophet for Retail Demand Fo...,7
7,red_thread_deep_001,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,8
8,red_thread_medium_002,THREAD: Beyond ARIMA and Prophet: What are you...,9
9,red_thread_deep_009,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...,10


---
# Crowding / Diversity Search

Unfiltered semantic search often returns results dominated by one source type — the SemanticSearch results above were all Reddit chunks. **Crowding** ensures diversity by retrieving top-k results per group using the `DataObjectSearchServiceAsyncClient` to run filtered semantic searches concurrently via `asyncio.gather`.

This pattern is especially valuable for RAG pipelines where you want context from multiple perspectives: community discussions (Reddit), meeting transcripts (Zoom), and technical documentation (PDF).

## Crowd by Source Type

Retrieve top-2 results per source type (Reddit, Zoom, PDF) for balanced cross-type representation. Compare to the SemanticSearch result earlier — the same kind of query now returns PDF documentation alongside Reddit discussions and Zoom transcripts.

In [23]:
query = 'What methods handle seasonal patterns in demand forecasting?'
fields = ['chunk_id', 'source_type', 'source_uri', 'text_content']

results = await crowded_search(
    COMBINED, query,
    crowd_field='source_type',
    crowd_values=['reddit', 'zoom', 'pdf'],
    top_k_per_value=2,
    fields=fields,
)

for source_type, hits in results.items():
    print(f'{source_type}: {len(hits)} results')

print(f'\nCrowded search (by source_type): "{query}"\n')
crowded_df(results, ['chunk_id', 'source_type', 'text_content'])

reddit: 2 results
zoom: 2 results
pdf: 2 results

Crowded search (by source_type): "What methods handle seasonal patterns in demand forecasting?"



,group,group_rank,chunk_id,source_type,text_content
0,reddit,1,red_thread_deep_000,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
1,reddit,2,red_thread_deep_011,reddit,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...
2,zoom,1,zoom_meeting_model_planning_000,zoom,"Sarah (Lead): Alright everyone, thanks for joi..."
3,zoom,2,zoom_meeting_model_planning_001,zoom,[Summary: The team debates Prophet versus a gr...
4,pdf,1,pdf_90bbc35dad_c19,pdf,# What's next\n\n• Try the following tutorials...
5,pdf,2,pdf_90bbc35dad_c16,pdf,# Custom holiday limitations\n\n• Custom holid...


## Crowd by Source URI

The `crowded_search` function works with **any** metadata field. Crowd by `source_uri` to ensure diversity across individual documents.

In [24]:
query = 'How accurate is ARIMA for retail forecasting?'
fields = ['chunk_id', 'source_type', 'source_uri', 'text_content']

# Get unique Reddit source URIs
reddit_results = list(search_client.query_data_objects(
    vectorsearch_v1beta.QueryDataObjectsRequest(
        parent=COMBINED,
        filter={'source_type': {'$eq': 'reddit'}},
        output_fields=vectorsearch_v1beta.OutputFields(data_fields=['source_uri']),
    )
))
uris = list(set(r.data['source_uri'] for r in reddit_results))
print(f'Crowding across {len(uris)} Reddit source URIs\n')

results = await crowded_search(
    COMBINED, query,
    crowd_field='source_uri',
    crowd_values=uris,
    top_k_per_value=1,
    fields=fields,
)

for uri, hits in results.items():
    short = uri.split('/')[-1]
    print(f'{short}: {len(hits)} result')

print(f'\nCrowded by source_uri: "{query}"\n')
crowded_df(results, ['chunk_id', 'source_type', 'source_uri', 'text_content'])

Crowding across 3 Reddit source URIs

thread_medium.json: 1 result
thread_shallow.json: 1 result
thread_deep.json: 1 result

Crowded by source_uri: "How accurate is ARIMA for retail forecasting?"



,group,group_rank,chunk_id,source_type,source_uri,text_content
0,gs://statmike-mlops-349915/applied-genai-solut...,1,red_thread_medium_004,reddit,gs://statmike-mlops-349915/applied-genai-solut...,THREAD: Beyond ARIMA and Prophet: What are you...
1,gs://statmike-mlops-349915/applied-genai-solut...,1,red_thread_shallow_000,reddit,gs://statmike-mlops-349915/applied-genai-solut...,THREAD: ARIMA vs. Prophet for Retail Demand Fo...
2,gs://statmike-mlops-349915/applied-genai-solut...,1,red_thread_deep_000,reddit,gs://statmike-mlops-349915/applied-genai-solut...,THREAD: ARIMA vs. Prophet vs. LightGBM for Dem...


---
# Key Takeaways

| Method | Strengths | Weaknesses | Best For |
|--------|-----------|------------|----------|
| **Query** | Fast metadata filtering, no index needed | No relevance ranking | Browsing, faceted navigation |
| **SemanticSearch** | Understands intent, finds synonyms | May miss exact keywords | Natural language questions |
| **TextSearch** | Precise keyword matching | No semantic understanding | Specific terms, acronyms, names |
| **VectorSearch (dense)** | Same as Semantic but with external models | Requires manual embedding | External embedding models |
| **VectorSearch (sparse)** | Statistical term importance via BM25 | Requires trained vocabulary | Term-weighted keyword search |
| **VectorSearch (dense+sparse)** | Combined vector scoring via RRF | Requires both embeddings | Hybrid dense+sparse retrieval |
| **Hybrid RRF (2-way)** | Best of semantic + keyword | Requires weight tuning | Production search |
| **Hybrid RRF (3-way)** | Three complementary signals | More API channels | Maximum recall |
| **VertexRanker** | Most accurate reranking (cross-encoder) | Not yet supported in VS2 | Quality-critical applications (future) |
| **Crowding** | Ensures diversity across groups | More API calls (async mitigates) | Multi-source RAG pipelines |

**RRF weight tuning tips:**
- Start with equal weights `[1, 1]` or `[1, 1, 1]` as a baseline
- Increase semantic weight when users ask natural language questions
- Increase text/sparse weights when users search for specific terms
- 3-way RRF adds BM25's statistical term importance as a third signal — the IDF component surfaces terms that are discriminating across the corpus

**VectorSearch `oneof` constraint:**
- `VectorSearch` uses a `oneof` for `vector` (dense) and `sparse_vector` — you can supply one or the other per search channel, not both
- To combine dense + sparse, use `batch_search_data_objects` with two `VectorSearch` channels fused via RRF

**VertexRanker vs RRF:**
- VertexRanker is a model-based cross-encoder — it reads each (query, document) pair and scores relevance directly. More accurate but slower.
- RRF is purely statistical — it fuses ranked lists by position without examining content. Fast and robust.
- VertexRanker is defined in the `vectorsearch_v1beta` API but not yet implemented by the backend. Use RRF for now.

**ANN vs kNN (SearchHint):**
- By default, VS2 auto-selects an ANN index if one exists, falling back to brute-force kNN
- Use `SearchHint(use_knn=True)` on `SemanticSearch` or `VectorSearch` to force exact kNN (useful for evaluation/ground truth)
- `TextSearch` does not support `SearchHint`
- At small scale (117 DataObjects), ANN and kNN produce identical results; ANN provides latency benefits at thousands+ of DataObjects
- See step 10 (`manage_ann_index.py`, `inspect_ann_index.ipynb`) for index management and recall benchmarking

**BM25 sparse embeddings:**
- BM25 sparse vectors capture term importance (IDF) and document length normalization — different from VS2's native TextSearch
- Built from `text_content + enrichment tags` (Approach B), which normalizes concepts across source types
- Vocabulary is stored in BigQuery for analysis and reconstruction (see step 9 maintenance)

**Manual embedding dimension matching:**
- When using `VectorSearch` with manually generated dense embeddings, you must set `output_dimensionality` to match the collection schema (768 in this project)
- `gemini-embedding-001` natively produces 3072 dims; VS2 auto-embed handles this internally for `SemanticSearch`